synapse/handlers/e2e_keys.py

```diff
@@ -608,6 +610,279 @@ def upload_signing_keys_for_user(self, user_id, keys):
+            target_device_id, signature), and a map of users to devices to failure
+            reasons
+        """
+        signature_list = []
+        failures = {}
+        if not signatures:
+            return signature_list, failures
+
+        try:
+            # get our self-signing key to verify the signatures
+            self_signing_key, self_signing_key_id, self_signing_verify_key = yield self._get_e2e_cross_signing_verify_key(
+                user_id, "self_signing"
+            )
+
+            # get our master key, since it may be signed
+            master_key, master_key_id, master_verify_key = yield self._get_e2e_cross_signing_verify_key(
+                user_id, "master"
+            )
+
+            # fetch our stored devices.  This is used to 1. verify
+            # signatures on the master key, and 2. to can compare with what
+            # was sent if the device was signed
+            devices = yield self.store.get_e2e_device_keys([(user_id, None)])
+
+            if user_id not in devices:
+                raise SynapseError(404, "No device keys found", Codes.NOT_FOUND)
+
+            devices = devices[user_id]
+        except SynapseError as e:
+            failures[user_id] = {
+                device: _exception_to_failure(e) for device in signatures.keys()
+            }
+            return signature_list, failures
+
+        for device_id, device in signatures.items():
+            try:
+                if "signatures" not in device or user_id not in device["signatures"]:
+                    # no signature was sent
+                    raise SynapseError(
+                        400, "Invalid signature", Codes.INVALID_SIGNATURE
+                    )
+
+                if device_id == master_verify_key.version:
+                    # we have master key signed by devices: for each
+                    # device that signed, check the signature.  Since
+                    # the "failures" property in the response only has
+                    # granularity up to the signed device, either all
+                    # of the signatures on the master key succeed, or
+                    # all fail.  So loop over the signatures and add
+                    # them to a separate signature list.  If everything
+                    # works out, then add them all to the main
+                    # signature list.  (In practice, we're likely to
+                    # only have only one signature anyways.)
+                    master_key_signature_list = []
+                    sigs = device["signatures"]
+                    for signing_key_id, signature in sigs[user_id].items():
+                        alg, signing_device_id = signing_key_id.split(":", 1)
+                        if (
+                            signing_device_id not in devices
+                            or signing_key_id
+                            not in devices[signing_device_id]["keys"]["keys"]
```

Comment:
ugh, `["keys"]["keys"]` is horrible, as is the `stored_device = devices[device_id]["keys"]` below. Any chance I can get you to take on a preliminary refactor of `get_e2e_device_keys` to return something slightly saner?

Reply:
I can do so, if you have suggestions for how to fix it.  It looks like `get_e2e_device_keys` is only used here and in `query_local_devices`.  There, it needs `device_info["keys"]` and `device_info["device_display_name"]` and modifies `device_info["keys"]`to add the display name.  Should I just make `query_local_devices` in change of adding in the device display name, and then have `devices[signing_device_id]["keys"]` become just `devices[signing_device_id]`?

Reply:
that sounds sensible, I think. I'd love it if it could be a preliminary PR rather than rolled into this one, though.